universal-application-tool-0.0.1/app/repository/VersionRepository.java

```diff
@@ -218,16 +226,56 @@ public boolean isDraft(Program program) {
   private BlockDefinition updateQuestionVersions(BlockDefinition block) {
     BlockDefinition.Builder updatedBlock =
         block.toBuilder().setProgramQuestionDefinitions(ImmutableList.of());
+    // Update questions contained in this block.
     for (ProgramQuestionDefinition question : block.programQuestionDefinitions()) {
       Optional<Question> updatedQuestion = getLatestVersionOfQuestion(question.id());
       LOG.trace(
           "Updating question ID {} to new ID {}.", question.id(), updatedQuestion.orElseThrow().id);
       updatedBlock.addQuestion(
           ProgramQuestionDefinition.create(updatedQuestion.orElseThrow().getQuestionDefinition()));
     }
+    // Update questions referenced in this block's predicate(s)
+    if (block.visibilityPredicate().isPresent()) {
+      PredicateDefinition oldPredicate = block.visibilityPredicate().get();
+      updatedBlock.setVisibilityPredicate(
+          PredicateDefinition.create(
+              updatePredicateNode(oldPredicate.rootNode()), oldPredicate.action()));
+    }
+    if (block.optionalPredicate().isPresent()) {
+      PredicateDefinition oldPredicate = block.optionalPredicate().get();
+      updatedBlock.setOptionalPredicate(
+          Optional.of(
+              PredicateDefinition.create(
+                  updatePredicateNode(oldPredicate.rootNode()), oldPredicate.action())));
+    }
     return updatedBlock.build();
   }
 
+  // Update the referenced question IDs in all leaf nodes. Since nodes are immutable, we
+  // recursively recreate the tree with updated leaf nodes.
+  @VisibleForTesting
```

Comment:
Praise: ohh, I like this annotation. Thanks for adding it.

Reply:
Thanks! It comes in handy when there are complicated methods you'd like to test!